src/videojs-hls.js

```diff
@@ -739,28 +797,37 @@ videojs.Hls.prototype.fillBuffer = function(seekToTime) {
     return;
   }
 
-  // find the next segment to download
-  if (typeof seekToTime === 'number') {
-    mediaIndex = this.playlists.getMediaIndexForTime_(seekToTime);
-  } else if (currentBuffered && currentBuffered.length) {
-    mediaIndex = this.playlists.getMediaIndexForTime_(currentBuffered.end(0));
-    bufferedTime = Math.max(0, currentBuffered.end(0) - currentTime);
-  } else {
-    mediaIndex = this.playlists.getMediaIndexForTime_(this.tech_.currentTime());
+  if (mediaIndex === undefined) {
+    if (currentBuffered && currentBuffered.length) {
+      mediaIndex = this.playlists.getMediaIndexForTime_(currentBuffered.end(0));
+      bufferedTime = Math.max(0, currentBuffered.end(0) - currentTime);
+      currentBufferedEnd = currentBuffered.end(0);
+
+      // if there is plenty of content in the buffer and we're not
+      // seeking, relax for awhile
+      if (bufferedTime >= videojs.Hls.GOAL_BUFFER_LENGTH) {
+        return;
+      }
+    } else {
+      mediaIndex = this.playlists.getMediaIndexForTime_(this.tech_.currentTime());
+    }
   }
   segment = this.playlists.media().segments[mediaIndex];
 
+  if (segment &&
+      this.lastSegmentUri_ &&
+      this.lastSegmentUri_.indexOf(segment.uri) !== -1) {
```

Comment:
If getMediaIndexForTime() didn't clamp, the instance variable wouldn't be necessary.